feat: Introduce changed-files GitHub action.

parasense opened the pull request:
Remarks:
* Add "changed-files" action to CI workflow
* Configure action to diff  "definitons/*/*.yaml" files
* Update shell script use changed-files as input.
* Added `on: path` [workflow filter pattern](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet) to ONLY run when bundle definitions are part of the push.
* Accordingly, removed redundant path checking REGEX from the script.

This should allow the repo to re-enable merge-commits.

Jira: hacbs-692

Signed-off-by: Jon Disnard <jdisnard@redhat.com>

davidmogar reviewed (commented):
As this is still a testing PR, one thing you can do is to modify the workflow so it doesn't require a push to main. That way you can fully test it within this PR without having to merge it.

johnbieren reviewed (commented):
I am a little confused as to why the docs show as part of this PR, but the changes to `tkn_push-bundles.sh` and `tekton_bundle_push.yaml` lgtm

parasense commented:
> I am a little confused as to why the docs show as part of this PR, but the changes to `tkn_push-bundles.sh` and `tekton_bundle_push.yaml` lgtm

Ah, Good catch!
That was probably caused by `git rebase origin/main` on my local branch, which is a habit of mine before pushing new changes. I ran another `git rebase origin/main` and it had merge conflicts, that I resolved. Hopefully that fixed it.

parasense commented:
> As this is still a testing PR, one thing you can do is to modify the workflow so it doesn't require a push to main. That way you can fully test it within this PR without having to merge it.

Good idea!
So I tried that, and it worked. 
I modified a few bundle definitions in trivial ways, and let the action run on this feature branch.
You can see the result here:
https://github.com/hacbs-release/release-bundles/runs/6902732290?check_suite_focus=true 
(This result might be garbage collected in the future)

The output log shows my test strings:
```
...
definitions/hello-release/hello-release.yaml was changed
definitions/hello-world/hello-world.yaml was changed
...
```
So I went ahead and just removed those extra testing lines from the script, and restored the branch to it's previous state to minimize the diff. 

PTAL

davidmogar commented:
Nicely done. I’ll go through it tomorrow and add my review then.

parasense commented:
> The docs appear here as the branch is not rebased (it has conflicts).
> 
> Lets not merge test stuff into main if that's not required though. The change is ok and I like that you (Jon) are using an action to detect changed files (if it works for your workflow), but let's keep it limited to this change whenever possible. If it's **actually necessary** to merge it to main, ask for a review and any of us will approve it.

This change is the first-part of a two-part effort to re-allow "merge commits" to the main branch. Previously our CI was silently failing due to the use of merge commits, rather than squash merge. The action makes it easier, but it's possible to have done the same kind of things directly in the workflow, but I'd rather reduce complexity. 

Once this PR is merged, we can update the script in another PR, and finish-up.

PTAL

davidmogar commented:
I think we can use this same PR (changing its scope) for applying all the necessary changes to get the CI fixed.

I’m also ok with merging this one and work in a separate one. In this case I’d update the PR description and commit as you removed some stuff. Also, if you go for this option, you can create a branch on top of this one to not get stuck until it’s reviewed and rebase it once the PR gets merged.

parasense commented:
> I think we can use this same PR (changing its scope) for applying all the necessary changes to get the CI fixed.
> 
> I’m also ok with merging this one and work in a separate one. In this case I’d update the PR description and commit as you removed some stuff. Also, if you go for this option, you can create a branch on top of this one to not get stuck until it’s reviewed and rebase it once the PR gets merged.

Okay, I've updated the PR with a few minimal changes to the script. 
I'll also update the PR to reflect the change.

Thanks!

johnbieren commented on .github/workflows/tekton_bundle_push.yaml:
```diff
@@ -28,9 +35,11 @@ jobs:
           password: ${{ env.REGISTRY_PASSWORD }}
           registry: ${{ env.IMAGE_REGISTRY }}
       - name: push tekton bundles
+        if: steps.changed-files.outputs.any_changed == 'true'
```
Isn't this line a duplication of the functionality on line 6?

johnbieren reviewed (approved):
The PR lgtm. I see a change in the `<>` in the README.md file, but I think that is just a rebase issue.

parasense commented:
> The PR lgtm. I see a change in the `<>` in the README.md file, but I think that is just a rebase issue.

Thank!
Yeah I'm not 100% clear why there was a rebase issue, but I *DO* rebase my feature branches overtop main on the regular. :man_shrugging:
